godot-core/src/builtin/vectors/vector_macros.rs

```diff
@@ -457,6 +457,61 @@ macro_rules! impl_vector_fns {
     }
 }
 
+pub(super) fn snap_one(mut value: i32, step: i32) -> i32 {
+    assert!(
+        value != i32::MIN || step >= 0,
+        "snapped() called on vector component i32::MIN with step component smaller then 0"
+    );
+    assert!(
+        value != i32::MAX || step <= 1,
+        "snapped() called on vector component i32::MAX with step component greater then 1"
+    );
+
+    if step != 0 {
+        // Can overflow if step < -1 and value is i32::MIN or step > 1 and value is i32::MAX.
+        let a = value + step / 2;
+
+        // Manual implement `a.div_floor(step)` since Rust's native method is still unstable, as of 1.79.0.
+
+        // Can overflow if value == i32::MIN and step == -1 because then a == i32::MIN.
+        let mut d = a / step;
+        // Can't overflow because if a == i32::MIN and step == -1, value == -2147483647.5 which is impossible.
+        let r = a % step;
+        if (r > 0 && step < 0) || (r < 0 && step > 0) {
+            // Can't overflow because if d == i32::MIN then a == i32::MIN and step == 1 and value == -2147483648.5 which is impossible.
+            d -= 1;
+        }
+
+        value = step * d;
+    }
+
+    value
+}
+
+/// Implements functions that are present only on integer vectors.
+macro_rules! impl_integer_vector_fns {
+    (
+        // Names of the components, for example `x, y`.
+        $($comp:ident),*
+    ) => {
+        /// A new vector with each component snapped to the closest multiple of the corresponding
+        /// component in `step`.
+        ///
+        /// # Panics
+        /// If any component of `self` is `i32::MIN` while the same component on `step` smaller then `0` or
+        /// if any component of `self` is `i32::MAX` while the same component on `step` greater then `1`.
```

Comment:
Maybe this could be worded in a way makes clear that panics only happen in under/overflow scenarios, not general usage and especially not 0 steps:

```suggestion
        /// # Panics
        /// On under- or overflow:
        /// - If any component of `self` is `i32::MIN` while the same component on `step` less than `0`.
        /// - If any component of `self` is `i32::MAX` while the same component on `step` greater than `1`.
```